cpp/src/arrow/flight/sql/example/sqlite_server.cc

```diff
@@ -627,28 +674,80 @@ class SQLiteFlightSqlServer::Impl {
     return DoGetSQLiteQuery(db_, query, SqlSchema::GetCrossReferenceSchema());
   }
 
-  Status ExecuteSql(const std::string& sql) {
+  Status ExecuteSql(const std::string& sql) { return ExecuteSql(db_, sql); }
+
+  Status ExecuteSql(sqlite3* db, const std::string& sql) {
     char* err_msg = nullptr;
-    int rc = sqlite3_exec(db_, sql.c_str(), nullptr, nullptr, &err_msg);
+    int rc = sqlite3_exec(db, sql.c_str(), nullptr, nullptr, &err_msg);
     if (rc != SQLITE_OK) {
       std::string error_msg;
       if (err_msg != nullptr) {
         error_msg = err_msg;
+        sqlite3_free(err_msg);
       }
-      sqlite3_free(err_msg);
-      return Status::ExecutionError(error_msg);
+      return Status::IOError(error_msg);
     }
+    if (err_msg) sqlite3_free(err_msg);
     return Status::OK();
   }
+
+  arrow::Result<ActionBeginTransactionResult> BeginTransaction(
+      const ServerCallContext& context, const ActionBeginTransactionRequest& request) {
+    std::string handle = GenerateRandomString();
+    sqlite3* new_db = nullptr;
+    if (sqlite3_open_v2(db_uri_.c_str(), &new_db,
+                        SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_URI,
+                        /*zVfs=*/nullptr) != SQLITE_OK) {
+      std::string error_message = "Can't open new connection: ";
+      if (new_db) {
+        error_message += sqlite3_errmsg(new_db);
+        sqlite3_close(new_db);
+      }
+      return Status::Invalid(error_message);
+    }
+
+    ARROW_RETURN_NOT_OK(ExecuteSql(new_db, "BEGIN TRANSACTION"));
+
+    open_transactions_[handle] = new_db;
+    return ActionBeginTransactionResult{std::move(handle)};
+  }
+
+  Status EndTransaction(const ServerCallContext& context,
+                        const ActionEndTransactionRequest& request) {
+    std::lock_guard<std::mutex> guard(mutex_);
+    auto it = open_transactions_.find(request.transaction_id);
+    if (it == open_transactions_.end()) {
+      return Status::KeyError("Unknown transaction ID: ", request.transaction_id);
+    }
+
+    Status status;
+    if (request.action == ActionEndTransactionRequest::kCommit) {
+      status = ExecuteSql(it->second, "COMMIT");
+    } else {
+      status = ExecuteSql(it->second, "ROLLBACK");
+    }
+    sqlite3_close(it->second);
```

Comment:
Not sure how efficient you want this to be, but you might release the lock around these lines (and call `open_transactions_.erase` before?).